shared/js/moz_intl.js

```diff
@@ -71,24 +71,32 @@ global.mozIntl = {
   DateTimeFormat: function(locales, options) {
     const resolvedOptions = Object.assign({}, options);
 
-    if (options.dayperiod) {
-      if (!('hour' in options)) {
+    if (resolvedOptions.dayperiod) {
+      if (resolvedOptions.hour === undefined) {
         resolvedOptions.hour = 'numeric';
       }
     }
 
-    if (resolvedOptions.hour &&
-        !options.hasOwnProperty('hour12')) {
+    if (resolvedOptions.hour !== undefined &&
+        resolvedOptions.hour12 === undefined) {
       resolvedOptions.hour12 = navigator.mozHour12;
     }
 
+    if (resolvedOptions.dayperiod === undefined &&
+        resolvedOptions.hour12 === true) {
+      resolvedOptions.dayperiod = true;
+    }
+
     var intlFormat = new Intl.DateTimeFormat(locales, resolvedOptions);
 
+    resolvedOptions.locale = intlFormat.resolvedOptions().locale;
+    resolvedOptions.hour12 = intlFormat.resolvedOptions().hour12;
+
     // This is needed for a workaround for bug 1208808
     // Remove when that bug is fixed
     var hourFormatter;
-    if (options.dayperiod === true || options.dayperiod === false &&
-        intlFormat.resolvedOptions().hour12 === true) {
+    if (resolvedOptions.dayperiod !== undefined &&
```

Comment:
Is there a way to combine all those `resolvedOptions.dayperiod` checks into a single if-else?